contracts/modules/token/TokenController.sol

```diff
@@ -391,24 +413,154 @@ contract TokenController is ITokenController, LockHandler, MasterAwareV2 {
+    return managerStakingPools[member].length > 0;
+  }
+
+  function getStakingPoolOwnershipOffer(
+    uint poolId
+  ) external override view returns (address proposedManager, uint deadline) {
+    return (
+      stakingPoolOwnershipOffers[poolId].proposedManager,
+      stakingPoolOwnershipOffers[poolId].deadline
+    );
+  }
+
+  /// Transfer ownership of all staking pools managed by a member to a new address. Used when switching membership.
+  ///
+  /// @param from  address of the member whose pools are being transferred
+  /// @param to    the new address of the member
+  function transferStakingPoolsOwnership(address from, address to) external override onlyInternal {
+
+    uint stakingPoolCount = managerStakingPools[from].length;
+
+    if (stakingPoolCount == 0) {
+      return;
+    }
+
+    while (stakingPoolCount > 0) {
+      // remove from old
+      uint poolId = managerStakingPools[from][stakingPoolCount - 1];
+      managerStakingPools[from].pop();
+
+      // add to new and update manager
+      managerStakingPools[to].push(poolId);
+      stakingPoolManagers[poolId] = to;
+
+      stakingPoolCount--;
+    }
+  }
+
+  function _assignStakingPoolManager(uint poolId, address manager) internal {
+
+    address previousManager = stakingPoolManagers[poolId];
+
+    // remove previous manager
+    if (previousManager != address(0)) {
+      uint managedPoolCount = managerStakingPools[previousManager].length;
+
+      // find staking pool id index and remove from previous manager's list
+      // on-chain iteration is expensive, but we don't expect to have many pools per manager
+      for (uint i = 0; i < managedPoolCount; i++) {
+        if (managerStakingPools[previousManager][i] == poolId) {
+          uint lastIndex = managedPoolCount - 1;
+          managerStakingPools[previousManager][i] = managerStakingPools[previousManager][lastIndex];
+          managerStakingPools[previousManager].pop();
+          break;
+        }
+      }
+    }
+
+    // add staking pool id to new manager's list
+    managerStakingPools[manager].push(poolId);
+    stakingPoolManagers[poolId] = manager;
+  }
+
+  /// Transfers the ownership of a staking pool to a new address
+  /// Used by PooledStaking during the migration
+  ///
+  /// @param poolId       id of the staking pool
+  /// @param manager      address of the new manager of the staking pool
+  function assignStakingPoolManager(uint poolId, address manager) external override onlyInternal {
+    _assignStakingPoolManager(poolId, manager);
+  }
+
+  /// Creates a ownership transfer offer for a staking pool
+  /// The offer can be accepted by the proposed manager before the deadline expires
+  ///
+  /// @param poolId           id of the staking pool
+  /// @param proposedManager  address of the proposed manager
+  /// @param deadline         timestamp after which the offer expires
+  function createStakingPoolOwnershipOffer(
```

Comment:
What's the reason for creating this offer -> accept flow instead of just transferring the ownership based on the current manager's decision?

Unclear from comments and original ticket

Reply:
Prevents denial of service attacks against members as you could create and transfer ownership of multiple staking pools to a particular address which would prevent them from voting in governance as the function wouldn't fit in a block. Fix is also applied when creating a staking pool (cannot assign a random address as a manager - msg.sender is used).

Reply:
Can we get some value range for what is an acceptable max for pool count, gas-wise, give we have those 2 nested loops here and the voting itself?

Doesn't need to be enforced but it would give us an idea.

Reply:
- There aren't any nested loops, it's just a single for loop
- Inside the loop we're reading to storage variables:
```
uint poolId = managerStakingPools[_of][i];
amount += stakingPoolNXMBalances[poolId].deposits;
```
This would place us in the range of hundreds (if not thousands) of staking pools - not a real concern given the staking pool ownership transfer has to be accepted.

Reply:
Awesome hundreds/thousands sounds great. Plenty of room for normal users to not be concerned about this value.

Only people pushing the system limits will actually worry about it.